src/aiodynamo/credentials.py

```diff
@@ -0,0 +1,312 @@
+    """
+    Ensures there can only ever be one "active" instance of the wrapped coroutine function.
+
+    Multiple concurrent calls will only result in a single call to the wrapped coroutine and
+    all callers will get the same result (or Exception).
+
+    Non-concurrent calls will behave as if the coroutine function is not wrapped in this decorator.
+    """
+    future = None
+
+    def reset_future(_):
+        nonlocal future
+        future = None
+
+    async def shielded(*args, **kwargs):
+        nonlocal future
+        future = asyncio.create_task(coro_func(*args, **kwargs))
+        future.add_done_callback(reset_future)
+        return await future
+
+    @wraps(coro_func)
+    async def wrapper(*args, **kwargs):
+        nonlocal future
+        if future is not None:
+            return await future
+        # This shield is needed in case the caller to this gets cancelled. Otherwise
+        # other "queued" callers will get a cancelled exception instead of the result.
+        return await asyncio.shield(shielded(*args, **kwargs))
+
+    return wrapper
+
+
+class Expiration(Enum):
+    not_expired = auto()
+    expires_soon = auto()
+    expired = auto()
+
+
+EXPIRES_SOON_THRESHOLD = datetime.timedelta(minutes=15)
+EXPIRED_THRESHOLD = datetime.timedelta(minutes=5)
+
+
+@dataclass(frozen=True)
+class Metadata:
+    key: Key
+    expires: datetime.datetime
+
+    def lives_longer_than(self, other: Metadata) -> bool:
+        return self.expires > other.expires
+
+    def check_expiration(self) -> Expiration:
+        now = datetime.datetime.now(datetime.timezone.utc)
+        if now >= self.expires:
+            return Expiration.expired
+        diff = self.expires - now
+        if diff < EXPIRED_THRESHOLD:
+            return Expiration.expired
+        if diff < EXPIRES_SOON_THRESHOLD:
+            return Expiration.expires_soon
+        return Expiration.not_expired
+
+
+@dataclass
+class MetadataCredentials(Credentials, metaclass=abc.ABCMeta):
+    @abc.abstractmethod
+    async def _fetch_metadata(self, http: HTTP) -> Metadata:
+        pass
+
+    async def fetch_with_retry(
+        self,
+        *,
+        http: HTTP,
+        max_attempts: int,
+        url: URL,
+        timeout: Timeout,
+        headers: Optional[Headers] = None
+    ) -> bytes:
+        for attempt in range(max_attempts):
+            try:
+                response = await http.get(url=url, headers=headers, timeout=timeout)
+            except:
+                logging.exception("GET failed")
+                continue
+            if response:
+                return response
+        raise TooManyRetries()
+
+    def __post_init__(self):
+        self._metadata: Optional[Metadata] = None
+
+    async def get_key(self, http: HTTP) -> Optional[Key]:
+        try:
+            await self._check_metadata(http)
+        except Disabled:
+            return None
+        return self._metadata and self._metadata.key
+
+    @highlander
```

Comment:
how does `highlander` work with multiple regions?
is there still only 1 metadata, or does it accidentally interlock credentials between regions?
(I've no idea at all how mutli-region-ness even works, maybe @michaelaw320 can elaborate)

Reply:
The client itself is region-locked (you explicitly pass it a region), so that's not a problem. The highlander is bound to an instance of `Credentials` and thus bound to a client (unless you share credentials between clients). If you need different credentials/regions/clients, you need multiple client instances with different credential instances.

Reply:
actually, after sleeping on it, I realized that the decorator is per-class, not per-instance. I'll need to change this.